File: Trees/Java/Print all nodes that don't have sibling.java
// Problem link: https://www.geeksforgeeks.org/problems/print-all-nodes-that-dont-have-sibling/1
// Time: O(nlogn)
// Space: O(n)
class Tree
{
    ArrayList<Integer> result;
    
    ArrayList<Integer> noSibling(Node node)
    {
        result = new ArrayList<>();
        find(node);
        
        if(result.size() == 0){
            result.add(-1);
            return result;
        }
        Collections.sort(result);
        return result;
    }
    
    private void find(Node node){
        if(node == null){
            return;
        }
        
        if(node.left == null || node.right == null){
            if(node.left != null){
                result.add(node.left.data);
            }
            else if(node.right != null){
                result.add(node.right.data);
            }
        }
        
        find(node.left);
        find(node.right);
    }
}

File: Trees/Python/Print all nodes that don't have sibling.py
# Problem link: https://www.geeksforgeeks.org/problems/print-all-nodes-that-dont-have-sibling/1
# Time: O(nlogn)
# Space: O(n)
def noSibling(root):
    result = []
    def find(node):
        if not node:
            return
        
        if not node.left or not node.right:
            if node.left:
                result.append(node.left.data)
            elif node.right:
                result.append(node.right.data)

        find(node.left)
        find(node.right)

    find(root)
    
    result.sort()
    return [-1] if len(result) == 0 else result
